File: lib/fetch_swagger_data_types.dart
enum TMethodType {
  get,
  post,
  put,
  delete,
  patch,
}

enum TMediaType {
  multipartFormData,
  applicationJson,
  textPlain,
  textJson,
  applicationJsonPlus,
}

enum TNumberType {
  integer,
  number,
}

//export type TPrimitiveType = TNumberType | "boolean" | "string"; // (string includes dates and files);
enum TPrimitiveType {
  integer,
  number,
  boolean,
  string,
}

//export type TDataType = TPrimitiveType | "array" | "object";
enum TDataType {
  integer,
  number,
  boolean,
  string,
  array,
  object,
}

class IPrimitiveProperty {
  final TPrimitiveType type;

  IPrimitiveProperty({required this.type});
}

class IArrayProperty {
  final TDataType type = TDataType.array;
  final IProperty? items;

  IArrayProperty({this.items});
}

class IRefProperty {
  final String
      ref; //`#/components/schemas/\${string}` | `#/components/schemas/\${string}.Enums.\${string}`;
  //type?: undefined;

  IRefProperty({required this.ref});
}

class IPropertyProperties {
  //[propertyName: string]: IProperty;
  final Map<String, IProperty> properties;

  IPropertyProperties({required this.properties});
}

class IObjectProperty {
  final TDataType type = TDataType.object;
  final IPropertyProperties? properties;
  final dynamic additionalProperties; //IProperty | true;

  IObjectProperty({this.properties, this.additionalProperties});
}

class IProperty {
  final TPrimitiveType? type;
  final String? ref;
  final bool? nullable;
  final List<dynamic>? enumeration;
  final List<IProperty>? oneOf;
  final List<IProperty>? anyOf;
  final List<IProperty>? allOf;
  final dynamic anyValue;

  IProperty({
    this.type,
    this.ref,
    this.nullable,
    this.enumeration,
    this.oneOf,
    this.anyOf,
    this.allOf,
    this.anyValue,
  });
}

class IParameter {
  final String name;
  final String
      inValue; //"query" | "header" | "path"; // in path example: /users/{userId}
  final IProperty? schema;
  final bool? required;

  IParameter({
    required this.name,
    required this.inValue,
    this.schema,
    this.required,
  });
}

class IResponse {
  final Map<String, dynamic> responses;
  IResponse({required this.responses});
}

class IRequestBody {
  final String? description;
  final Map<TMediaType, dynamic>? content;

  IRequestBody({this.description, this.content});
}

class IOpenApiJSON {
  final String openapi; // version
  final IInfo info;
  final List<IServer>? servers;
  final Map<String, dynamic> paths;
  final IComponents components;

  IOpenApiJSON({
    required this.openapi,
    required this.info,
    this.servers,
    required this.paths,
    required this.components,
  });
}

class IInfo {
  final String title;
  final String description;
  final String version;
  final IContact? contact;

  IInfo({
    required this.title,
    required this.description,
    required this.version,
    this.contact,
  });
}

class IContact {
  final String? name;

  IContact({this.name});
}

class IServer {
  final String url;
  final String description;

  IServer({required this.url, required this.description});
}

class IComponents {
  final Map<String, IProperty> schemas;

  IComponents({required this.schemas});
}

class TControllerInfo {
  final Map<String, List<TRouteInfo>> controllerInfo;

  TControllerInfo({required this.controllerInfo});
}

class TRouteInfo {
  final Map<String, dynamic> routeInfo;

  TRouteInfo({required this.routeInfo});
}

File: typescript/src/types/fetch-swagger-data.types.ts
export type TMethodType = "get" | "post" | "put" | "delete" | "patch";

export type TMediaType =
  | "multipart/form-data"
  | "application/json"
  | "text/plain"
  | "text/json"
  | "application/*+json";

export type TNumberType = "integer" | "number";

export type TPrimitiveType = TNumberType | "boolean" | "string"; // (string includes dates and files);

export type TDataType = TPrimitiveType | "array" | "object";

export type IPrimitiveProperty = {
  type: TPrimitiveType;
};

export type IArrayProperty = {
  type: "array";
  items?: IProperty;
};

export type IRefProperty = {
  $ref:
    | `#/components/schemas/${string}`
    | `#/components/schemas/${string}.Enums.${string}`;
  type?: undefined;
};

export type IPropertyProperties = {
  [propertyName: string]: IProperty;
};

export type IObjectProperty = {
  type: "object";
  properties?: IPropertyProperties;
  additionalProperties?: IProperty | true;
};

export type IProperty = (
  | IPrimitiveProperty
  | IArrayProperty
  | IRefProperty
  | IObjectProperty
  | {
      $ref: "#/components/schemas/System.TimeSpan"; // equal to string in typescript
      type?: undefined;
    }
) & {
  nullable?: boolean;
  enum?: (number | string)[]; // example-1
  oneOf?: IProperty[]; // match one of them (equal to | in typescript)
  anyOf?: IProperty[]; // array each element must be one of them (equal to (|)[] in typescript)
  allOf?: IProperty[]; //  takes an array of object and merge them to a single object (equal to & in typescript)
  AnyValue?: {}; // equal to any
};

// example-1:
// {
// type: string;
// enum: [test1, test2];
// }
// => result in typescript should be:
//  {
// property: test1 | test2
// }

export type IParameter = {
  name: string;
  in: "query" | "header" | "path"; // in path example: /users/{userId}
  schema?: IProperty;
  required?: boolean;
};

export type IResponse = {
  "200": {
    description?: string;
    content?: {
      [mediaType in TMediaType]?: {
        schema: IProperty;
      };
    };
  };
};

export type IRequestBody = IResponse["200"];

export type IOpenApiJSON = {
  openapi: string; // version
  info: {
    title: string;
    description: string;
    version: string;
    contact?: {
      name: string;
    };
  };
  servers?: {
    url: string;
    description: string;
  }[];
  paths: {
    [route: string]: {
      [methodType in TMethodType]?: {
        summary?: string;
        description?: string;
        parameters?: IParameter[];
        requestBody?: IRequestBody;
        responses: IResponse;
      };
    };
  };
  components: {
    schemas: {
      [dtoOrEnumName: string]: IProperty;
    };
  };
};

// ----------------------------------------------------------------------

export type TControllerInfo = {
  [controllerName: string]: TRouteInfo[];
};

export type TRouteInfo = {
  [route: string]: {
    hasResponse: boolean;
    hasBodyPayload: boolean;
    isArrayBodyPayload: boolean;
    hasQueryParams: boolean;
    hasPathParams: boolean;
    hasPageParamInBody: string[];
    hasPageParamInQuery: string[];
    methodType: TMethodType;
    mediaType: TMediaType;
    parametersInPath: IParameter[];
  };
};
